teps/0058-graceful-pipeline-run-termination.md

```diff
@@ -136,32 +149,59 @@ Consider including folks that also work outside the WGs or subproject.
 
 ### User Experience (optional)
 
-<!--
-Consideration about the user experience. Depending on the area of change,
-users may be task and pipeline editors, they may trigger task and pipeline
-runs or they may be responsible for monitoring the execution of runs,
-via CLI, dashboard or a monitoring system.
-
-Consider including folks that also work on CLI and dashboard.
--->
+Support for additional command or flags in `tkn` CLI should be considered.
 
 ### Performance (optional)
 
 No impact on performance.
 
 ## Design Details
 
-<!--
-This section should contain enough information that the specifics of your
-change are understandable.  This may include API specs (though not always
-required) or even code snippets.  If there's any ambiguity about HOW your
-proposal will be implemented, this is the place to discuss them.
-
-If it's helpful to include workflow diagrams or any other related images,
-add them under "/teps/images/". It's upto the TEP author to choose the name
-of the file, but general guidance is to include at least TEP number in the
-file name, for example, "/teps/images/NNNN-workflow.jpg".
--->
+In this proposal we need to consider following cases:
+
+1. User sets `spec.status` to "PipelineRunStopped" in `PipelineRun` with running tasks, but `finally` section is empty.
+
+    In this case a graceful termination would ack the same as a pipeline run cancellation.
+    
+    - `spec.status` in all running `TaskRun` is patched to "TaskRunCancelled"
+    - `spec.status` in all running `Run` is patched to "RunCancelled"
+    - `PipelineRun` condition after graceful termination is:
+      ```yaml
+      { type: Succeeded, Status: False, Reason: Cancelled } 
+      ```
+
+2. User sets `spec.status` to "PipelineRunStopped" in `PipelineRun` with running tasks and non-empty `finally` section.
+
+    In this case a graceful termination cancels all running task runs and waits for final tasks to be processed.
+    
+    - `spec.status` in all running `TaskRun` is patched to "TaskRunCancelled"
+    - `spec.status` in all running `Run` is patched to "RunCancelled"
+    - `PipelineRun` condition after graceful termination is:
+      ```yaml
+      { type: Succeeded, Status: Unknown, Reason: PipelineRunStopping } 
+      ```    
+
+3. User sets `spec.status` to "PipelineRunStopping" in `PipelineRun` with running tasks.
+
+    In this case a graceful termination waits for all tasks to be completed.
+    
+    - `PipelineRun` condition after graceful termination is:
+      ```yaml
+      { type: Succeeded, Status: Unknown, Reason: PipelineRunStopping } 
```

Comment:
one more thought: i'd like to also push back on using unknown here - am in understanding correctly that the intention here is that even after all the tasks in the pipelinerun have stopped, the status would be "unknown"? We use "unknown" to mean "currently running" (https://github.com/tektoncd/pipeline/blob/main/docs/pipelineruns.md#monitoring-execution-status) so it wouldn't make sense to use that here - I think the only thing that would make sense is to set the status to False, as it in it failed to run completely

(im also wondering if there are more details in your use cases that might explain why you want to use unknown here)

Reply:
Nice catch! Status `Unknown` means that the pipeline is running.
The condition can be `{ type: Succeeded, Status: Unknown, Reason: PipelineRunStopping }` right after the user requested cancellation, but ultimately it should be something like `{ type: Succeeded, Status: False, Reason: CancelledByUser }` or so.

Reply:
My interpretation here is the status will be `unknown` since some `taskRuns` are still executing and pipeline is going to schedule `finally` after the existing `taskRuns` finish executing. And eventually the status would change to `False` and the reason stays the same.

Reply:
hey @rafalbigaj please add the PipelineRun condition values in the end of the run for all the use cases. 🙏

Reply:
That was my intention, but probably I missed to write it up. The `Unknown` status, after termination, was expected to indicate that processing is still in progress. When the progressing finishes the status would be `False`. I'm going to make it more verbose in the proposal.